### PR TITLE
libnetwork: write ServFail if DNS reply msg is bad

### PR DESCRIPTION
**Pull request by corhere**
- Relates to #46928 

If the resolver's DNSBackend returns a name that cannot be marshaled into a well-formed DNS message, the resolver will only discover this when it attempts to write the reply message and it fails with an error. No reply message is sent, leaving the client to wait out its timeout and the user in the dark about what went wrong.

**- What I did**

Make a best-effort attempt to inform the client of the failure.

**- How I did it**

When writing the intended reply message fails, retry once with a ServFail response to inform the client and user that the DNS query was not resolved due to a problem with to the resolver, not the network.

**- How to verify it**

New unit test.

**- Description for the changelog**

- The container DNS resolver will send ServFail replies instead of failing silently in more situations.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by corhere** [comment]
cc @robmry

File: libnetwork/resolver.go [hunk 1]
```diff
@@ -381,6 +381,15 @@ func (r *Resolver) serveDNS(w dns.ResponseWriter, query *dns.Msg) {
 			r.log(ctx).WithError(err).Errorf("[resolver] failed to write response")
 			span.RecordError(err)
 			span.SetStatus(codes.Error, "WriteMsg failed")
+			// Make a best-effort attempt to send a failure response to the
+			// client so it doesn't have to wait for a timeout if the failure
+			// has to do with the content of msg rather than the connection.
+			if msg.Rcode != dns.RcodeServerFailure {
+				if err := w.WriteMsg(new(dns.Msg).SetRcode(query, dns.RcodeServerFailure)); err != nil {
+					r.log(ctx).WithError(err).Errorf("[resolver] writing ServFail response also failed")
```
Comment (thaJeztah) [inline libnetwork/resolver.go:389]: 🙈 

```suggestion
					r.log(ctx).WithError(err).Error("[resolver] writing ServFail response also failed")
```

**Review by akerouanton — APPROVED** [review]
Except the broken tests, this LGTM. Thanks!

**Review by robmry — APPROVED** [review]
LGTM too - and the test failure is just duff input ...

```
--- a/libnetwork/resolver_unix_test.go
+++ b/libnetwork/resolver_unix_test.go
@@ -64 +64 @@ func TestDNSIPQuery(t *testing.T) {
-       names := []string{"name1", "NaMe1"}
+       names := []string{"name1.", "NaMe1."}
@@ -87 +87 @@ func TestDNSIPQuery(t *testing.T) {
-       q.SetQuestion("name1", dns.TypeMX)
+       q.SetQuestion("name1.", dns.TypeMX)
@@ -99 +99 @@ func TestDNSIPQuery(t *testing.T) {
-       q.SetQuestion("nonexistent", dns.TypeMX)
+       q.SetQuestion("nonexistent.", dns.TypeMX)
```

**Comment by corhere** [comment]
Should be all fixed now. Serves me right for not running all the package tests locally.

**Review by thaJeztah — APPROVED** [review]
LGTM if CI is ✅

**Comment by thaJeztah** [comment]
Windows.. never change. Thanks for the informative errors;

```
=== Failed
=== FAIL: github.com/docker/docker/integration/container TestWindowsDevices/process/class/5B45201D-F2F2-4F3B-85BB-30FF1F953599_mounted (1.48s)
    devices_windows_test.go:107: assertion failed: error is not nil: Error response from daemon: failed to create task for container: failed to create shim task: hcs::CreateComputeSystem 6ca2659c76ba9a58eb85e692ac17b108190d989e5a2f5b23e53d879f2215e8cb: The parameter is incorrect.: unknown
```

```
=== FAIL: github.com/docker/docker/integration/container TestLogs/driver_json-file/tty/stdout_and_stderr (1.50s)
    logs_test.go:134: assertion failed: error is not nil: Error response from daemon: hcs::CreateComputeSystem 202c0491b6a8ccd311fc30340026b2b36165716ae343b77ed7da30103d9008e6: Element not found.
```

**Pull request merged**